### PR TITLE
Created caching memory allocator

### PR DESCRIPTION
**Pull request by oierlauzi**
The caching memory allocator implemented here is based on PyTorch's approach:
https://zdevito.github.io/2022/08/04/cuda-caching-allocator.html

The problem with cudaMalloc/cudaFree/cudaHostMalloc/cudaHostFree is that they implicitly **synchronize** all CUDA streams, which is detrimental for performance. The benefit of a caching allocator is that calls to the former functions are reduced significantly.

With this PR, the current compute API is fully implemented in CUDA. However, writing tests (specially unit tests for the caching allocator) is still pending.

### CODE REVIEW

**Comment by oierlauzi** [comment]
Something odd happens with Windows path. Even if I quote it, it still splits it into separated tokens

**Comment by MartinSalinas98** [comment]
I think CMake always expectes forward slashes to be used, so the Windows-like path would need to be parsed into Unix-like.

**Comment by oierlauzi** [comment]
Maybe skip windows for now and later add it back?

**Comment by MartinSalinas98** [comment]
> Maybe skip windows for now and later add it back?

If this attempt does not work, i'll skip it and open this research in a new PR to work on it in parallel

File: src/compute/cuda/allocator/cuda_memory_block_pool.inl [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+{
+
+static void* offset(void *ptr, std::ptrdiff_t count)
+{
+    return static_cast<void*>(static_cast<char*>(ptr)+count);
+}
+
+
+
+
```
Comment (MartinSalinas98) [inline src/compute/cuda/allocator/cuda_memory_block_pool.inl:51]: Remove extra empty lines

File: src/compute/cuda/allocator/cuda_memory_block_pool.inl [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+inline
+bool cuda_memory_block_context::is_free() const noexcept
+{
+    return m_free;
+}
+
+
+
+
```
Comment (MartinSalinas98) [inline src/compute/cuda/allocator/cuda_memory_block_pool.inl:103]: Extra empty lines

File: src/compute/cuda/allocator/cuda_memory_block_pool.inl [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+inline
+bool check_links(cuda_memory_block_pool::iterator ite) noexcept
+{
+    return check_backward_link(ite) && check_forward_link(ite);
+}
+
+
+
+
```
Comment (MartinSalinas98) [inline src/compute/cuda/allocator/cuda_memory_block_pool.inl:201]: Extra lines. Are they intended?

**Review by MartinSalinas98 — CHANGES_REQUESTED** [review]

**Comment by MartinSalinas98** [comment]
I wasn't able to see anything I didn't like as far as I understand the code. We can do one of two things:
- I approve the PR and gets merged
- We have a meet call to do a code review steo by step so I can understand the code better and provide more meaningful feedback

**Comment by oierlauzi** [comment]
We can take a watch this afternoon

**Review by MartinSalinas98 — CHANGES_REQUESTED** [review]

**Review by MartinSalinas98 — DISMISSED** [review]

**Pull request merged**